### PR TITLE
[Paddle][CUDAGraph] 175B GPT-3 Hybrid-Parallel Training with CUDAGraph

### PR DESCRIPTION
**Pull request by eee4017**
# Description

In this PR, we introduce support for CUDAGraph in TE-PaddlePaddle. The primary issue with CUDAGraph is managing branching. Such as when enabling `weight_cache`, specific operations are required only in the first microbatch, but branching within CUDAGraph is undesirable.

### Solution 1: Utilizing Multiple Graphs (TE-PyTorch Solution)

TE-PyTorch addresses branching by recording separate graphs: one for the true branch and another for the false branch. This method necessitates maintaining distinct CUDA graphs for each microbatch. However, this solution can become complex, especially with the Pipeline Parallelism mechanism, as the number of required graphs doubles with each branching point. Managing 2^N graphs for N branchings can lead to significant challenges.

### Solution 2: Reordering Kernel Sequences

To simplify the process, we reorder the kernel sequences outside the computational graph, keeping the branching outside the graph's scope. This approach avoids the complexity of managing multiple graphs and ensures efficient execution within CUDAGraph.

## Changes Introduced in this PR

- Support for random seed updates (`set_rng_state`) using the set parameter mechanism in Paddle.
- Support for `amax_and_scale_update_inplace` with the set parameter mechanism in Paddle. This kernel is manually issued, preserving the legacy kernel for now while we explore other solutions.
- Reordering FP8 weight cache before the PP pipe to exclude it from the CUDAGraph.

## Type of Change

- [ ] Documentation change (changes only to documentation, either a fix or new content)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infrastructure/Build change
- [ ] Code refactor

# Checklist

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

# Testing

Although we have not yet implemented unit tests, integration tests have been validated. The following parallelism configurations are supported:

1. BF16: PP / TP / PP+TP / PP+TP+SP
2. FP8: PP / TP / PP+TP / PP+TP+SP

## Performance Testing on GPT-3 1.3B on 4 H100 GPUs

|   Model Configuration  | CUDAGraph | Loss | Samples/sec | Speedup |
|:----------------------:|:-----------:|:----:|:-----------:|:-------:|
| 1.3B, BF16, PP2+TP2    | N         | 9.44 | 28.43       | 0.00%   |
| 1.3B, BF16, PP2+TP2    | Y         | 9.44 | 55.32       | 94.54%  |
| 1.3B, BF16, PP2+TP2+SP | N         | 9.44 | 24.38       | 0.00%   |
| 1.3B, BF16, PP2+TP2+SP | Y         | 9.44 | 50.12       | 105.54% |
| 1.3B, BF16, PP4        | N         | 9.59 | 53.81       | 0.00%   |
| 1.3B, BF16, PP4        | Y         | 9.59 | 68.80       | 27.87%  |
| 1.3B, BF16, TP4        | N         | 9.69 | 14.54       | 0.00%   |
| 1.3B, BF16, TP4        | Y         | 9.69 | 37.76       | 159.65% |
| 1.3B, FP8, PP2+TP2     | N         | 9.44 | 18.42       | 0.00%   |
| 1.3B, FP8, PP2+TP2     | Y         | 9.44 | 55.84       | 203.06% |
| 1.3B, FP8, PP2+TP2+SP  | N         | 9.47 | 16.26       | 0.00%   |
| 1.3B, FP8, PP2+TP2+SP  | Y         | 9.47 | 51.83       | 218.71% |
| 1.3B, FP8, PP4         | N         | 9.59 | 40.73       | 0.00%   |
| 1.3B, FP8, PP4         | Y         | 9.59 | 76.31       | 87.37%  |
## Performance Testing on GPT-3 175B with 64 H100 GPUs
We achieved convergence within the first 1024 steps on an 8-node cluster.

image

+ Performance Gain
  - BF16: Improved from 11.1 to 11.5, achieving a speedup of 1.03x.
  - FP8: Improved from 13.8 to 16.3, achieving a speedup of 1.18x.

### CODE REVIEW

File: transformer_engine/paddle/csrc/custom_ops.cu [hunk 1]
```diff
@@ -581,11 +582,43 @@ std::vector<paddle::Tensor> te_rmsnorm_bwd(const paddle::Tensor &dz, const paddl
...
+    auto seed_offset = gen_cuda->IncrementOffset(rng_elts_per_thread);
+    params.As<std::pair<int64_t, int64_t>>(1) = seed_offset;
+  };
+
+  phi::backends::gpu::CUDAGraphNodeLauncher::cudaKernelCallback_t cudaKernelCallback =
+      [=](unsigned int id) {
+        void *functionPtr = reinterpret_cast<void *>(&set_rng_state);
+        cudaFunction_t cudaFunc;
+        PADDLE_ENFORCE_GPU_SUCCESS(cudaGetFuncBySymbol(&cudaFunc, functionPtr));
+        set_rng_state<<<1, 1, 0, stream>>>(id, seed_offset, rng_state_p);
```
Comment (eee4017) [inline transformer_engine/paddle/csrc/custom_ops.cu:615]: solved with set parameter for the `set_rng_state` at parameter  `seed_offset`

File: transformer_engine/paddle/layer/base.py [hunk 2]
```diff
@@ -455,21 +477,78 @@ def get_fp8_weights_scratchpad(
...
+                                    FP8FwdTensors.GEMM1_WEIGHT
+                                    if i == 1
+                                    else FP8FwdTensors.GEMM2_WEIGHT
+                                ),
+                                fp8_dtype_forward,
+                                out=weight_fp8,
+                            )
+
+            cast_callback(0 if is_first_microbatch else 1)
+            register_pp_fwd_begin_hook(cast_callback)
```
Comment (eee4017) [inline transformer_engine/paddle/layer/base.py:552]: The casting kernel is recorder to the begining of a PP Pipe

File: transformer_engine/paddle/csrc/custom_ops.cu [hunk 3]
```diff
@@ -8,8 +8,9 @@
 #include <map>
 #include <vector>
 
+#include "../common.h"
```
Comment (jeng1220) [inline transformer_engine/paddle/csrc/custom_ops.cu:11]: The ".." is at transformer_engine/paddle, but there is no common.h.

File: transformer_engine/paddle/fp8.py [hunk 4]
```diff
@@ -235,25 +246,44 @@ def amax_and_scale_update(
     fp8_meta: Dict[str, Any],
     fwd_update: bool,
     update_weight_scale_inv: bool = True,
+    current_step_id_tensor: Optional[Any] = None,
```
Comment (timmoon10) [inline transformer_engine/paddle/fp8.py:249]:
```suggestion
    current_step_id_tensor: Optional[paddle.Tensor] = None,
```
It's not intuitive to me that this is a CPU `int` tensor, so it may be worth documenting or adding a comment.

File: transformer_engine/paddle/csrc/custom_ops.cu [hunk 5]
```diff
@@ -1277,6 +1344,58 @@ void amax_and_scale_update_inplace(paddle::Tensor &amax_history,  // NOLINT
...
+    amax = amax_history.slice(0, 1);
+  }
+
+  const auto rolled_amax_history = amax_history.roll({-1}, {0});
+
+  auto amax_history_numel = amax_history.numel();
+  auto amax_numel = amax.numel();
+  size_t num_blocks = (amax_history_numel + BLOCK_SIZE - 1) / BLOCK_SIZE;
+
+  auto parameterSetter = [current_step_id_tensor](phi::backends::gpu::CUDAKernelParams &params) {
```
Comment (timmoon10) [inline transformer_engine/paddle/csrc/custom_ops.cu:1370]: I'm not sure how expensive it is to copy a Paddle tensor, but no harm capturing-by-reference to avoid triggering the copy-constructor:
```suggestion
  auto parameterSetter = [&current_step_id_tensor](phi::backends::gpu::CUDAKernelParams &params) {
```

File: transformer_engine/paddle/fp8_buffer.py [hunk 6]
```diff
@@ -179,14 +202,17 @@ def wait(self) -> None:
...
         return out
 
     def from_numpy(self, buffer: Dict[str, np.array]) -> None:
         """Set buffer values from numpy arrays"""
         for k, v in buffer.items():
-            self._data[k] = [paddle.to_tensor(arr) for arr in v]
+            self._global_amax[k] = [paddle.to_tensor(arr) for arr in v]
+
+    def enable_cudagraph(self):
+        self._use_cudagraph = True
```
Comment (timmoon10) [inline transformer_engine/paddle/fp8_buffer.py:215]: Should we support the case where we disable CUDA graphs?
```suggestion
    def enable_cudagraph(self, enable: bool = True) -> None:
        self._use_cudagraph = enable
```
For example, a user might want to pretrain with CUDA graphs and do finetuning without. This may be something to defer until later if it will require significant work.

File: transformer_engine/paddle/distributed.py [hunk 7]
```diff
@@ -13,6 +13,11 @@
 import paddle.distributed.fleet.base.topology as tp
 from paddle.distributed.fleet.meta_parallel import get_rng_state_tracker
 from paddle.distributed.fleet.layers.mpu import mp_ops
+from paddle.distributed.fleet.meta_parallel import (
+    PipelineParallel,
+    PipelineParallelMicroStepLocations,
+    register_global_pipeline_parallel_hook,
```
Comment (timmoon10) [inline transformer_engine/paddle/distributed.py:19]: I don't see `register_global_pipeline_parallel_hook` in the Paddle `develop` branch or recent releases: https://github.com/PaddlePaddle/Paddle/blob/81dbca14788bdf68de2b3f2b87f8a16aa2e84342/python/paddle/distributed/fleet/meta_parallel/__init__.py

File: transformer_engine/paddle/distributed.py [hunk 8]
```diff
@@ -54,6 +59,19 @@ def get_tp_group_and_world_size(
     return model_parallel_group, world_size
 
 
+def is_pp_enabled() -> int:
+    if not (paddle.distributed.is_initialized()):
+        return 1
+
+    return tp._HYBRID_PARALLEL_GROUP.get_pipe_parallel_world_size()
```
Comment (timmoon10) [inline transformer_engine/paddle/distributed.py:66]: Based on the name and usage in `TransformerEngineBaseLayer`:
```suggestion
def is_pp_enabled() -> bool:
    if not paddle.distributed.is_initialized():
        return False
    return tp._HYBRID_PARALLEL_GROUP.get_pipe_parallel_world_size() > 1
```
If the implementation is intended, the name should be changed to something like `get_pipeline_parallel_size` to avoid confusion.

File: transformer_engine/paddle/layer/transformer.py [hunk 9]
```diff
@@ -154,6 +157,9 @@ def __init__(
         self.tensor_parallel = self.tp_size > 1
         self.sequence_parallel = self.tensor_parallel and sequence_parallel
         self.hidden_dropout_rng_state_name = hidden_dropout_rng_state_name
+        self.use_cudagraph = use_cudagraph
```
Comment (timmoon10) [inline transformer_engine/paddle/layer/transformer.py:160]: I don't think the `use_cudagraph` attribute is used anywhere:
```suggestion
```

File: transformer_engine/paddle/layer/transformer.py [hunk 10]
```diff
@@ -139,6 +141,7 @@ def __init__(
         attention_dropout_rng_state_name: str = "local_seed",
         hidden_dropout_rng_state_name: str = "global_seed",
         backend: str = "transformer_engine",
+        use_cudagraph: bool = True,
```
Comment (timmoon10) [inline transformer_engine/paddle/layer/transformer.py:144]: CUDA graphs are delicate, so I prefer disabling them by default:
```suggestion
        use_cudagraph: bool = False,
```
Also, this API might run into problems if a user tries to initialize some layers with `use_cudagraph=True` and other layers with `use_cudagraph=False`. The `get_global_fp8_state().enable_cudagraph()` API is somewhat better at conveying that CUDA graph support is all-or-nothing.

File: transformer_engine/paddle/csrc/custom_ops.cu [hunk 11]
```diff
@@ -581,11 +582,43 @@ std::vector<paddle::Tensor> te_rmsnorm_bwd(const paddle::Tensor &dz, const paddl
   return {dx, dgamma};
 }
 
-__global__ void set_rng_state(std::pair<uint64_t, uint64_t> seed_offset, int64_t *rng_state_ptr) {
+__global__ void set_rng_state(
+    unsigned int identifier,  // This is used to relate kernel to cudaGraph nodes
```
[the remark this answers is not in the document]
Comment (eee4017) [reply transformer_engine/paddle/csrc/custom_ops.cu:586]: This parameter is essential. For more details, please refer to the comments [here](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/backends/gpu/cuda/cuda_graph.h#L121). This identifier plays a crucial role in linking the CUDA kernel to its corresponding CUDA graph node. After launching a kernel, each kernel launch is tagged with a unique identifier. This tagging allows us to use `cudaGraphKernelNodeGetParams` to retrieve information from the CUDA graph about the kernel. The mechanism is designed this way because we only obtain the kernel pointer after using `cudaGraphKernelNodeGetParams` to parse the recorded CUDA graph and identify the kernel. It would be beneficial if CUDA could design a more user-friendly method for using `cudaGraphKernelNodeSetParams` and `cudaGraphKernelNodeGetParams`. This functionality is implemented in [PR #60516](https://github.com/PaddlePaddle/Paddle/pull/60516). I will add more comments. Additionally, we might consider adding a C++17 `[[maybe_unused]]` attribute here.

File: transformer_engine/paddle/csrc/custom_ops.cu [hunk 12]
```diff
@@ -1260,6 +1305,28 @@ void te_scaled_upper_triang_masked_softmax_backward(paddle::Tensor &output_grads
       softmax_results.stream());
 }
 
+__global__ void UpdateFP8MetaKernel(
```
[the remark this answers is not in the document]
Comment (eee4017) [reply transformer_engine/paddle/csrc/custom_ops.cu:1308]: Regarding this issue, what I really need is a custom kernel that can utilize the identifier/parameterSetter mechanism since we need the first parameter of the kernel to be the identifier. Therefore, I reverted to the legacy implementation (see [this PR](https://github.com/NVIDIA/TransformerEngine/pull/593/files#diff-0d7172c528739ab6b144f09788506478439b3748d259466544dbea1a7f0ee18bL1192)). The identifier/parameterSetter mechanism cannot be used if we employ NVTE kernels, as I prefer to keep the complex identifier/parameterSetter mechanism confined to this CU file. Do you have any suggestions on how to achieve this better?

File: transformer_engine/paddle/layer/transformer.py (hunk 10 above)
Comment (eee4017) [reply transformer_engine/paddle/layer/transformer.py:144]: You're right; let's use only `get_global_fp8_state().enable_cudagraph()` to enable cudaGraph globally.

File: transformer_engine/paddle/fp8_buffer.py (hunk 6 above)
Comment (eee4017) [reply transformer_engine/paddle/fp8_buffer.py:215]: From my past experience, enabling cudaGraph and then disabling it causes unexpected segmentation faults. We will investigate this issue further in the future if this feature is truly needed. For now, we should maintain the enable-only approach.

File: transformer_engine/paddle/csrc/custom_ops.cu (hunk 3 above)
Comment (eee4017) [reply transformer_engine/paddle/csrc/custom_ops.cu:11]: I may have missed this during the rebase process. I will restored to `#include "common/common.h"`. Thank you for pointing this out.

File: transformer_engine/paddle/distributed.py (hunk 8 above)
Comment (eee4017) [reply transformer_engine/paddle/distributed.py:66]: Thank you for pointing out the mistake. I have corrected it accordingly.

File: transformer_engine/paddle/layer/transformer.py (hunk 9 above)
Comment (eee4017) [reply transformer_engine/paddle/layer/transformer.py:160]: Removed.

File: transformer_engine/paddle/distributed.py (hunk 7 above)
Comment (timmoon10) [reply transformer_engine/paddle/distributed.py:18]: We should guard the import then:
```python
try:
    from paddle.distributed.fleet.meta_parallel import register_global_pipeline_parallel_hook
except ImportError:
    register_global_pipeline_parallel_hook = None
```
We should also update `register_pp_fwd_begin_hook`:
```python
def register_pp_fwd_begin_hook(forward_begin_hook):
    if register_global_pipeline_parallel_hook is not None:
        register_global_pipeline_parallel_hook(
            PipelineParallelMicroStepLocations.FORWARD_BEGIN,
            forward_begin_hook,
        )
```

File: transformer_engine/paddle/distributed.py [hunk 13]
```diff
@@ -13,10 +13,15 @@
 import paddle.distributed.fleet.base.topology as tp
 from paddle.distributed.fleet.meta_parallel import get_rng_state_tracker
 from paddle.distributed.fleet.layers.mpu import mp_ops
-from paddle.distributed.fleet.meta_parallel import (
-    PipelineParallelMicroStepLocations,
-    register_global_pipeline_parallel_hook,
-)
+
+try:
+    # This feature is only supported in the latest version of Paddle.
```
Comment (timmoon10) [inline transformer_engine/paddle/distributed.py:18]:
```suggestion
    # This feature is not supported as of Paddle 2.6.
```

**Review by timmoon10 — APPROVED** [review]

**Comment by timmoon10** [comment]
The CI is failing while building since it can't find glog:
```
In file included from /opt/transformerengine/transformer_engine/paddle/csrc/custom_ops.cu:13:
/usr/local/lib/python3.10/dist-packages/paddle/include/paddle/phi/backends/gpu/cuda/cuda_graph.h:30:10: fatal error: glog/logging.h: No such file or directory
   30 | #include "glog/logging.h"
```

**Review by timmoon10 — APPROVED** [review]
